File: src/tuner/domain/firmware_capabilities.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True, frozen=True)
class FirmwareCapabilities:
    source: str
    serial_protocol_version: int | None = None
    blocking_factor: int | None = None
    table_blocking_factor: int | None = None
    live_data_size: int | None = None
    supports_board_capabilities_channel: bool = False
    supports_spi_flash_health_channel: bool = False
    supports_runtime_status_a: bool = False
    experimental_u16p2: bool = False

    def runtime_trust_summary(self) -> str:
        """Human-readable summary of which runtime channel groups are trustworthy."""
        parts: list[str] = []
        if self.supports_runtime_status_a:
            parts.append("runtimeStatusA: trusted")
        else:
            parts.append("runtimeStatusA: uncertain (not advertised)")
        if self.supports_board_capabilities_channel:
            parts.append("boardCapabilities: available")
        if self.supports_spi_flash_health_channel:
            parts.append("spiFlashHealth: available")
        if self.experimental_u16p2:
            parts.append("U16P2: experimental")
        if self.live_data_size is not None:
            parts.append(f"live_data_size: {self.live_data_size}")
        return "; ".join(parts) if parts else "no capability detail"

    def uncertain_channel_groups(self) -> frozenset[str]:
        """Channel groups that should be treated as uncertain given these capabilities.

        Callers can use this set to suppress or flag runtime evidence for channels
        that the firmware did not advertise as present/reliable.
        """
        uncertain: set[str] = set()
        if not self.supports_runtime_status_a:
            uncertain.add("runtimeStatusA")
        if not self.supports_board_capabilities_channel:
            uncertain.add("boardCapabilities")
        if not self.supports_spi_flash_health_channel:
            uncertain.add("spiFlashHealth")
        return frozenset(uncertain)

File: cpp/src/firmware_capabilities.cpp
// SPDX-License-Identifier: MIT
#include "tuner_core/firmware_capabilities.hpp"
#include <cstdio>

namespace tuner_core::firmware_capabilities {

std::string Capabilities::runtime_trust_summary() const {
    std::string parts;
    auto append = [&](const char* s) {
        if (!parts.empty()) parts += "; ";
        parts += s;
    };
    if (supports_runtime_status_a)
        append("runtimeStatusA: trusted");
    else
        append("runtimeStatusA: uncertain (not advertised)");
    if (supports_board_capabilities_channel)
        append("boardCapabilities: available");
    if (supports_spi_flash_health_channel)
        append("spiFlashHealth: available");
    if (experimental_u16p2)
        append("U16P2: experimental");
    if (live_data_size.has_value()) {
        char buf[64];
        std::snprintf(buf, sizeof(buf), "live_data_size: %d", *live_data_size);
        parts += "; ";
        parts += buf;
    }
    return parts.empty() ? "no capability detail" : parts;
}

std::set<std::string> Capabilities::uncertain_channel_groups() const {
    std::set<std::string> uncertain;
    if (!supports_runtime_status_a) uncertain.insert("runtimeStatusA");
    if (!supports_board_capabilities_channel) uncertain.insert("boardCapabilities");
    if (!supports_spi_flash_health_channel) uncertain.insert("spiFlashHealth");
    return uncertain;
}

}  // namespace tuner_core::firmware_capabilities
